src/tests/test_taskrunner.py

```diff
@@ -0,0 +1,123 @@
+        for case in expected:
+            present = False
+            for row in result["stdout"].splitlines():
+                if row.strip().startswith(case):
+                    present = True
+                    break
+            self.assertTrue(present, "task %r not present in task listing" % case)
+
+    def test_commands_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("ping", task_list)
+        expected = {
+            'task': 'ping',
+            'task_args': [],
+            'task_kwargs': {},
+            'result': 'pong',
+            'rc': SUCCESS_RC}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_args_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("echo:zulu", task_list)
+        expected = {
+            'rc': 0,
+            'result': 'received: zulu',
+            'task': 'echo',
+            'task_args': ['zulu'],
+            'task_kwargs': {}}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_args_and_kwargs_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("echo:zulu,gamma,a=b", task_list)
+        expected = {
+            'rc': 0,
+            'result': "received: zulu with args: ('gamma',) and kwargs: {'a': 'b'}",
+            'task': 'echo',
+            'task_args': ['zulu', 'gamma'],
+            'task_kwargs': {'a': 'b'}}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_whitespace_can_be_called(self):
+        task_list = tr.generate_task_list()
+        # result_map = tr.exec_task("echo:msg,'echo hello world'", task_list) # this is what we give bash
+        result_map = tr.exec_task("echo:msg,echo hello world", task_list) # this is what we see after bash
+        expected = {
+            'task': 'echo',
+            'task_args': ['msg', 'echo hello world'],
+            'task_kwargs': {},
+            'result': "received: msg with args: ('echo hello world',) and kwargs: {}",
+            'rc': SUCCESS_RC}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_special_chars_can_be_called_without_quoting(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("echo:hello-world!@$%^&*()-;?/", task_list) # '!@' makes bash cry and doesn't work without quoting
+        expected = {
+            'rc': 0,
+            'result': 'received: hello-world!@$%^&*()-;?/',
+            'task': 'echo',
+            'task_args': ['hello-world!@$%^&*()-;?/'],
+            'task_kwargs': {}}
+        self.assertEqual(expected, result_map)
+
+    def test_quoted_commands_can_be_called(self):
+        task_list = tr.generate_task_list()
+        # result_map = tr.exec_task("echo:'hello\, world'", task_list) # this is what we give bash
+        result_map = tr.exec_task("echo:hello\\, world", task_list) # this is what we see after bash
```

Comment:
the pair of what we give bash and what we see doesn't check out for me; a single quoted `'hello\, world'` should preserve a single slash

Reply:
you're right, test updated